### PR TITLE
Count digits

### PR DESCRIPTION
**Pull request by GaiSamuel**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [ ] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by Zeinab15 — COMMENTED** [review]
Great work @GaiSamuel ! The function is clear and easy to follow, with good documentation. It handles errors effectively, and the tests cover many scenarios, making sure the function works as expected in different cases.

File: solutions/count_digits.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    - count_digits: Counts the number of digits in a given number.
+
+Created on Wednesday, 1st January, 2025.
+@author: Gai Samuel
+"""
+
+
+def count_digits(number):
+    """
+    The count_digits function counts the number of integers in a given number.
```
Comment (TekaMesfinAbel) [inline solutions/count_digits.py:15]: The phrase "counts the number of integers" is misleading. It should be revised to "counts the number of digits in the provided number."

File: solutions/count_digits.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    Parameters:
+        number: an integer or float number whose digits are to be counted.
+
+    Returns:
+        An integer representing the number of digits in the input number.
+        Note that the negative sign is not counted as a digit.
+
+    Raises:
+        ValueError if the input is empty.
```
Comment (TekaMesfinAbel) [inline solutions/count_digits.py:25]: he input type check and empty check can be clarified. For instance, an empty string raises a ValueError, but other empty-like values (e.g., None) raise TypeError. The docstring should reflect this.

File: solutions/tests/test_count_digits.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    def test_invalid_string_input(self):
+        """It raises a value error for an invalid string input"""
+        with self.assertRaises(ValueError):
+            count_digits("Gai")
+
+    def test_invalid_input(self):
+        """It raises a type error for an invalid input"""
+        with self.assertRaises(TypeError):
+            count_digits([1, 2, 3])
```
Comment (TekaMesfinAbel) [inline solutions/tests/test_count_digits.py:56]: Consider adding tests for other invalid types, such as None, dict, or bool, to ensure complete coverage.

File: solutions/tests/test_count_digits.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    def test_string_input(self):
+        """It returns the number of digits in a string."""
+        self.assertEqual(count_digits("15"), 2)
+
+    def test_empty_input(self):
+        """It raises a value error for an empty input"""
+        with self.assertRaises(ValueError):
+            count_digits(
+                (""),
+            )
```
Comment (TekaMesfinAbel) [inline solutions/tests/test_count_digits.py:46]: Remove the extra parentheses around the argument.

**Review by TekaMesfinAbel — COMMENTED** [review]
The implementation of count_digits and its test suite is well-structured and covers a variety of edge cases, including handling negative numbers, floats, strings, and invalid inputs. However, there are areas where improvements can be made for better clarity, robustness, and adherence to best practices. Specifically, the docstrings could be refined for better readability, and some redundant checks can be simplified or made more explicit. The test cases are comprehensive but would benefit from minor adjustments to improve their clarity and accuracy.

File: solutions/count_digits.py (hunk 1 above)
Comment (GaiSamuel) [reply solutions/count_digits.py:15]: Oops! I must've rushed with that one! Thanks for pointing it out.

**Review by Zeinab15 — COMMENTED** [review]
I have a few suggestions

File: solutions/tests/test_count_digits.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    def test_invalid_input(self):
+        """It raises a type error for an invalid input"""
+        with self.assertRaises(TypeError):
+            count_digits([1, 2, 3])
+            count_digits(None)
+            count_digits({1, 2, 3})
+            count_digits(True)
+            count_digits(False)
+            count_digits({"name": "Gai"})
```
Comment (Zeinab15) [inline solutions/tests/test_count_digits.py:59]: Following the checklist, you’ve made several improvement in the last test:

One Assertion per Test Case: You’ve made sure that each test case contains only one assertion, which is a best practice to isolate and clearly define the expected behavior.


Helpful Test Names: The test functions now have meaningful names.

Descriptive Docstrings: Each test includes a clear and concise docstring, explaining what the test case is verifying, which helps to understand the test’s intent at a glance.

File: solutions/count_digits.py (hunk 1 above)
Comment (GaiSamuel) [reply solutions/count_digits.py]: fixed

File: solutions/tests/test_count_digits.py (hunk 3 above)
Comment (GaiSamuel) [reply solutions/tests/test_count_digits.py:54]: Done

File: solutions/tests/test_count_digits.py (hunk 4 above)
Comment (GaiSamuel) [reply solutions/tests/test_count_digits.py]: Done

File: solutions/count_digits.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+Module contents:
+    - count_digits: Counts the number of digits in a given number.
+
+Created on Wednesday, 1st January, 2025.
+@author: Gai Samuel
+"""
+
+
+def count_digits(number):
```
[the remark this answers is not in the document]
Comment (GaiSamuel) [reply solutions/count_digits.py]: Done.

**Review by Zeinab15 — APPROVED** [review]
Well done!

File: solutions/tests/test_count_digits.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        """It raises a value error for an invalid string input"""
+        with self.assertRaises(ValueError):
+            count_digits("Gai")
+
+    def test_invalid_input(self):
+        """It raises a type error for an invalid input"""
+        with self.assertRaises(TypeError):
+            count_digits([1, 2, 3])
+            count_digits(None)
+            count_digits({1, 2, 3})
```
Comment (Mohamed-Elnageeb) [inline solutions/tests/test_count_digits.py:56]: How do these work? Like how many errors does this raise?

Comment (GaiSamuel) [reply solutions/tests/test_count_digits.py:56]: Those are all invalid inputs, and they all raise type errors. So, it was better to put all of them under one test.

Comment (Zeinab15) [reply solutions/tests/test_count_digits.py:56]: @GaiSamuel  I also suggest splitting the tests and providing a separate docstring for each input type. This will make it easier to identify which specific input caused a failure, improving test clarity, debugging efficiency, and overall readability.

Comment (Mohamed-Elnageeb) [reply solutions/tests/test_count_digits.py:56]: yeah I understood your point now, but i tried it in my pc andactually this syntax isn't doing what you expect, you can check it using the debugger. you will find that it checks the first line and then immediately goes to the last one indicating that it only checks the first line and then returns. So as zeinab mentioned you have to split them.

Comment (GaiSamuel) [reply solutions/tests/test_count_digits.py:56]: I noticed that too on Zeinab's suggestion and the issue was when testing the Booleans. I've split everything now to have their own tests.

Comment (Mohamed-Elnageeb) [reply solutions/tests/test_count_digits.py:56]: ok push it and i will approve rn

**Review by TekaMesfinAbel — APPROVED** [review]
Well done.
